cache.go

```diff
@@ -0,0 +1,276 @@
+package skewer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+)
+
+// Cache stores a list of known skus, possibly fetched with a provided client
+type Cache struct {
+	location string
+	filter   string
+	client   client
+	data     []SKU
+}
+
+// CacheOption describes functional options to customize the listing behavior of the cache.
+type CacheOption func(c *Cache) error
```

Comment:
One thing I run into a lot when I write variadic options is that I often start by making the fields of the input private b/c they belong to the struct I want to configure and be in a known, good state after configuration. I think this has undesirable consequences. For starters, it makes it impossible for a consumer of the API to create their own `CacheOption`. I mean, they _can_ make one, but they have no access to manipulate the configuration struct. Furthermore, the intent is misleading. You show me a public type, but give me nothing but behaviors :[.

I started doing this a slightly different way. Instead of the `CacheOption` taking the final object you want to construct, a `Cache`, have it take a `CacheConfig` structure which has public fields. Finally, if you don't intend on having fields accessible on the `Cache`, which makes sense in this case, you can alway just return a `Cacher` interface from `NewCache`, or perhaps `NewCacher`, and make `Cache` a `cache`.

Just my 2¢.

Reply:
I agree and struggled to find a good fit here. I was torn between making the api harder to abusing and adding more error conditions. I also was tempted to add background refresh in the future in which case doing something like changing the location or filter manually would have weird semantics (what happens before refresh? the data is not valid for what you configured anymore)

concretely, the main example I don't like exposing on an option struct is the client. if you get rid of the functional style for the clients you're left with the location option, which might as well be a function parameter instead of a functional opt. Before I wrote it like this I had 4 different constructors with slightly different behavior, and location was passed as a string arg.

If I keep it like this I might do what you suggested with the interface types

Reply:
sorta related to why the arguments all have context currently -- if I add either refresh or allow the cache to lazily retrieve results from the initial iterator, instead of draining it, we will require a context. The first one I think different signatures may be better but for a lazy iterator I think we'd required the context to traverse values/pages

Reply:
I'm not suggesting to get rid of the variadic options. I really like them. I'm suggesting the following.
```go
type CacheConfig struct {
	Location string
	Filter   string
	Data     []SKU
}

type Cache struct {
	config   *CacheConfig
	client   client
}

// CacheOption describes functional options to customize the listing behavior of the cache.
type CacheOption func(c * CacheConfig) error
```

Reply:
I like this, except the tricky part is the one you didn't move to the struct :) 

https://github.com/Azure/skewer/pull/1/files#diff-56ffa6659747ba7a23d2b5597a4cda5dR66

there are 3 functional options for client with different underlying types, I'd be curious how you'd model that on a struct like this. filter is also derived from location so really the only options are data/nil (data also doesn't pass in as a functional option, I guess I could replace NewStaticCache() with WithData() or something like that, but I don't know that I prefer that.

Reply:
maybe the config contains `client client` and the functional options operator on that, with the setters working as is? sorta the same problem as original

Reply:
The only difference is the client would be immutable after the initial configuration.

Reply:
hmm, is something like this what you are thinking? I'm a little stuck around the behavior for `client client`

https://diffy.org/diff/oe4brutjlkx24aim1bt8adcxr

Reply:
the client is the main reason to use functional options at all, but is also the hardest piece to model on a config struct

Reply:
Yep, that diff is pretty solid. 

One comment. perhaps, `func WithClient(client client) Option` should be `func withClient(client client) Option`.

Reply:
oooooh, I get the reasoning. the functional options can be used to modify the Cache after instantiation. That's not what we want.